### ISSUE TITLE
Update fails on Mac Running Lion

### ISSUE THREAD

**Issue by RobK88** [yitsushi/totp-cli#61]
I was able to install totp-cli on my Mac running Lion using Macporst without any issue.

I was also able to successfully run `sudo totp-cli update`.  But after the update, totp-cli no longer runs.
See below:

```
bash-3.2$ sudo port clean totp-cli
--->  Cleaning totp-cli
bash-3.2$ 
bash-3.2$ sudo port install totp-cli
--->  Computing dependencies for totp-cli
--->  Fetching distfiles for totp-cli
--->  Verifying checksums for totp-cli
--->  Extracting totp-cli
--->  Configuring totp-cli
--->  Building totp-cli
--->  Staging totp-cli into destroot                     
--->  Installing totp-cli @1.1.17_0
--->  Activating totp-cli @1.1.17_0
--->  Cleaning totp-cli
--->  Scanning binaries for linking errors
--->  No broken files found.                             
--->  No broken ports found.
bash-3.2$ 
bash-3.2$ totp-cli version
totp-cli v1.1.17 (darwin/amd64)
bash-3.2$ 
bash-3.2$ sudo totp-cli update
Check...
 -> Download...
 -> Extract...
Now you have a fresh new totp-cli \o/
bash-3.2$ 
bash-3.2$ totp-cli version
dyld: Symbol not found: _clock_gettime
  Referenced from: /opt/local/bin/totp-cli
  Expected in: flat namespace

Trace/BPT trap: 5
bash-3.2$
```

Lion only runs an old vesion of Go.  Perhaps that is the issue.

```
bash-3.2$ go version
go version go1.17.13 darwin/amd64
bash-3.2$
```

**Comment by yitsushi** [comment]
Can you please check what `file` says about the file?
```
file $(which totp-cli)
```

I downloaded the `tar.gz`, extracted it, it says
```
totp-cli: Mach-O 64-bit x86_64 executable
```

But I can't test it as I do not own a mac machine anymore.

---

For now I would say ignore the `update command` install it with
```
go install github.com/yitsushi/totp-cli@latest
```

**Comment by yitsushi** [comment]
> Lion only runs an old vesion of Go. Perhaps that is the issue.

It downloads a pre-built binary, so it should be an issue at all. It does not even need go to be installed to use `go update` or download the binary from the releases page.

**Comment by yitsushi** [comment]
Interesting.

> go1.16.2 needs an OS greater than MacOS 10.11.6 "OS X El Capitan" according to the golang download documents.
> 
> Source: https://stackoverflow.com/questions/66741383/dyld-symbol-not-found-clock-gettime-after-installing-golang

Based on that, it is possible even if go is not installed, the built symbol map of the binary is not something Lion likes. Based on that, I'm not even sure how to resolve this.

**Comment by yitsushi** [comment]
If nothing works, you can try:
```
git clone https://github.com/yitsushi/totp-cli
cd totp-cli
# edit go.mod if required (change go 1.19)
# I don't think I used anything that's 1.19 specific.
go build .
./totp-cli version
```

**Comment by RobK88** [comment]
Here is the output you requested:

```
bash-3.2$ file $(which totp-cli)
/opt/local/bin/totp-cli: Mach-O 64-bit executable x86_64
```

And here are the dependencies:
```

bash-3.2$ otool -L `which totp-cli`
/opt/local/bin/totp-cli:
	/opt/local/lib/libMacportsLegacySupport.dylib (compatibility version 1.0.0, current version 1.0.7)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 159.1.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 635.21.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 55148.6.0)
bash-3.2$
```

I will try building from source using git and will get back to you.

**Comment by RobK88** [comment]
I tried building from source using git.  I get the same error:

```
bash-3.2$ git clone https://github.com/yitsushi/totp-cli
Cloning into 'totp-cli'...
remote: Enumerating objects: 662, done.
remote: Counting objects: 100% (210/210), done.
remote: Compressing objects: 100% (113/113), done.
remote: Total 662 (delta 127), reused 127 (delta 92), pack-reused 452
Receiving objects: 100% (662/662), 4.04 MiB | 3.55 MiB/s, done.
Resolving deltas: 100% (344/344), done.

bash-3.2$ cd totp-cli/

bash-3.2$ go build .
go: downloading github.com/yitsushi/go-commander v1.1.0
go: downloading github.com/kardianos/osext v0.0.0-20190222173326-2bc1f35cddc0
go: downloading github.com/yitsushi/github-release-check v1.0.1
go: downloading gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
go: downloading golang.org/x/term v0.0.0-20220411215600-e5f449aeb171
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading golang.org/x/sys v0.0.0-20220412211240-33da011f77ad
# github.com/yitsushi/totp-cli
/opt/local/lib/go/pkg/tool/darwin_amd64/link: running /usr/bin/clang failed: exit status 1
ld: warning: can't parse dwarf compilation unit info in /tmp/go-link-3789674970/go.o
Undefined symbols for architecture x86_64:
  "_clock_gettime", referenced from:
      _runtime.walltime_trampoline.abi0 in go.o
  "_getentropy", referenced from:
      _internal/syscall/unix.libc_getentropy_trampoline.abi0 in go.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

It looks like it is using clang to link the object code.  The clang compiler that comes with Lion is old.  I have installed newer version of clang such as clang-9.0 and clang-15 using Macports.  Is there a way to use that linker instead?

**Comment by RobK88** [comment]
After I updated the port file (see attached) for totp-cli, I also tried building the latest version (v1.2.5) of totp-cli using Macports.   AT least with Macports, I can easily tell it to use a newer version of clang.

The build also failed but with a different error:

```
:info:build internal/terminal/terminal.go:10:2: cannot find package "golang.org/x/term" in any of:
:info:build     /opt/local/lib/go/src/golang.org/x/term (from $GOROOT)
:info:build     /opt/local/var/macports/build/_Users_grinch_Macports_ports_security_totp-cli/totp-cli/work/gopath/src/golang.org/x/term (from $GOPATH)
:info:build Command failed:  cd "/opt/local/var/macports/build/_Users_grinch_Macports_ports_security_totp-cli/totp-cli/work/gopath/src/github.com/yitsushi/totp-cli" && /opt/local/bin/go build 
:info:build Exit code: 1
:error:build Failed to build totp-cli: command execution failed
:debug:build Error code: CHILDSTATUS 3910 1
:debug:build Backtrace: command execution failed
:debug:build     while executing
:debug:build "system {*}$notty {*}$callback {*}$nice $fullcmdstring"
:debug:build     invoked from within
:debug:build "command_exec -callback portprogress::target_progress_callback build"
:debug:build     (procedure "portbuild::build_main" line 8)
:debug:build     invoked from within
:debug:build "$procedure $targetname"
:error:build See /opt/local/var/macports/logs/_Users_grinch_Macports_ports_security_totp-cli/totp-cli/main.log for details
```

Looks like there is a missing dependency in the Portfile.   Please take a look at the attached Portfile.  If you know what the missing dependency is, please let me know and I will try again.  
[Portfile.txt](https://github.com/yitsushi/totp-cli/files/10471162/Portfile.txt)

**Comment by RobK88** [comment]
P.S.  I created a ticket at Macports 
see https://trac.macports.org/ticket/66727#ticket

**Comment by RobK88** [comment]
I updated and fixed the Portfile for `totp-cli`.  I order for version 1.2.5 to build using Macports, a dependency 
 -- `golang/x/term` -- needed to be added to the Portfile.  When I have time I will submit a Pull Request to the Macports repo.

see 
https://trac.macports.org/ticket/66727#ticket

WIth the updated Portfile, I was able to build and run `totp-cli` v1.2.5 even on my old Mac running Lion!

```
Grinchs-Mac-Pro:~ grinch$ totp-cli version
totp-cli v1.2.5 (darwin/amd64)
```

**Comment by RobK88** [comment]
I have submitted a Pull Request to Macports for `totp-cli` version 1.2.5.  But it does not build on macOS11.  Pity!  Working on a fix.

**Comment by yitsushi** [comment]
```
/opt/local/lib/go/pkg/tool/darwin_amd64/link: running /usr/bin/clang failed: exit status 1
ld: warning: can't parse dwarf compilation unit info in /tmp/go-link-3789674970/go.o
Undefined symbols for architecture x86_64:
  "_clock_gettime", referenced from:
      _runtime.walltime_trampoline.abi0 in go.o
  "_getentropy", referenced from:
      _internal/syscall/unix.libc_getentropy_trampoline.abi0 in go.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

That looks something like an issue with the old version of the environment. As the Go download page suggests, Lion is not really supported by newer Go version (including 1.17). Fixing on Lion feels like a real hack and any changes that makes it working on Lion seems something that may not suitable for newer environment. But that's just my gut telling me things, I can't experiment with it as I don't have any Darwin systems in my house (only Linux and windows are available to me), tried to start a MacOS machine on AWS, but it  did not let me as it requires reserved dedicated hardware and my account has a limit of 0 (shame on them).

I put a `wonfix` label on this issue and close it. Discussion will not be locked, so updates are welcome and if I can do anything related do this, we can reopen the issue.

**Comment by RobK88** [comment]
Well `totp-cli` v1.2.5 does build and run just fine using Macports on Lion.  I just needed to update the Portfile and add a missing dependency.  

But unfortunately,  version 1.2.5 does not build on the latest version of macOS (Big Sur and later).  Pity.  I am hoping some at Macports can find the bug.  I suspect he problem is with the go compiler installed by Macports.

**Comment by RobK88** [comment]
I was able to fix the Macports Portfile so totp-cli v1.2.5 will build and run on newer versions of macOS.

The Portfile for totp-cli v1.2.5 will now build and run totp-cli not only on my old Mac running Lion but also on Big Sur and Monterey. I needed to add a missing dependency (golang/x/term) and change one of the dependencies (golang.org/x/sys) to the latest version!

P.S.  The reason why `totp-cli` builds and runs just fine on Lion is due to the Macports Legacy support library.
